src/number_theory/modular_forms/congruence_subgroups.lean

```diff
@@ -0,0 +1,234 @@
+/-
+Copyright (c) 2022 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+import linear_algebra.special_linear_group
+import data.zmod.basic
+import group_theory.subgroup.pointwise
+import group_theory.group_action.conj_act
+/-!
+# Congruence subgroups
+
+This defines congruence subgroups of `SL(2,ℤ)` such as `Γ(N)`, `Γ₀(N)` and `Γ₁(N)` for `N` a
+natural number.
+
+It also contains basic results about congruence subgroups.
+
+-/
+
+local notation `SL(` n `, ` R `)`:= matrix.special_linear_group (fin n) R
+
+local attribute [-instance] matrix.special_linear_group.has_coe_to_fun
+
+local prefix `↑ₘ`:1024 := @coe _ (matrix (fin 2) (fin 2) _) _
+
+open matrix.special_linear_group matrix
+
+/--The reduction mod `(N : ℕ)` map on `SL(2,ℤ)`. -/
+def SL_reduction_mod_hom (N : ℕ) : SL(2, ℤ) →* SL(2, (zmod N)) := map (int.cast_ring_hom (zmod N))
```

Comment:
I am not sure we want a `def` here. You can maybe have `variable (N : ℕ)` and use a local notation. The reason is that with a new definition you loose (at least part of) the API for `map`.

Reply:
Sure I made it notation, I just called it `SLMOD(N)` but if you have any preferences I can change it.